dataset/src/test/scala/frameless/sql/rules/SQLRulesSuite.scala

```diff
@@ -17,6 +17,27 @@ trait SQLRulesSuite extends TypedDatasetSuite with Matchers { self =>
     s"$tmpDir/${self.getClass.getName}"
   }
 
+  lazy val sparkFullVersion = {
+    val pos = classOf[Expression].getPackage.getSpecificationVersion
+    if (pos eq null) // DBR is always null
+      SparkSession.active.version // taking a running spark version string, hence lazy
+    else
+      pos
+  }
+
+  lazy val sparkVersion = {
+    sparkFullVersion.split('.').take(2).mkString(".")
+  }
+
+  /**
+   * Don't run this test on 3.2
+   */
+  def not3_2[T](thunk: => T): Any =
+    if (sparkVersion != "3.2")
+      thunk
+    else
+      ()
```

Comment:
I'm more in favor of back-porting the function into the FramelessInternals, these hacks are harder to maintain.

Reply:
no, I agree, just wanted a "correct" impl committing first.

Reply:
I'll let you finish it then first :D

Reply:
moreover having a divergent impl isn't great so it'll be another compat layer....

Reply:
Yea, so just a spark cross compilable function is good enough :+1: We just need to remove it once we drop Spark 3.2

Reply:
Ok, that last push has the goods.  There is a gotcha, in that there are probably other Expressions that don't implement foldable in 3.2 but do in later versions [searching for the obvious](https://issues.apache.org/jira/browse/SPARK-40380?jql=project%20%3D%20SPARK%20AND%20text%20~%20%22constant%20folding%22%20order%20by%20key%20desc) only shows [SPARK-40380](https://issues.apache.org/jira/browse/SPARK-40380) and [SPARK-39106](https://issues.apache.org/jira/browse/SPARK-39106) so we may be ok as I cover both.

In the case where there is another impl difference the worst that should happen now is it doesn't get pushed down.

Reply:
idk; we should not patch Spark bugs imo, I was only suggesting adding a cross compiling function that will properly work for Spark 3.3+ and will still be incorrect for the Spark 3.2, like here: https://github.com/typelevel/frameless/pull/721#issuecomment-1582729739